rerun_cpp/src/rerun/component_batch.hpp

```diff
@@ -1,94 +1,314 @@
 #pragma once
 
 #include <array>
+#include <utility>
 #include <vector>
 
-#include "data_cell.hpp"
 #include "result.hpp"
+#include "serialized_component_batch.hpp"
 
 namespace rerun {
+    /// The ComponentBatchAdaptor trait is responsible for mapping an input argument to a
+    /// ComponentBatch.
+    ///
+    /// There are default implementations for standard containers of components, as well as single
+    /// components.
+    ///
+    /// An adapter may choose to either produce a owned or borrowed component batch.
+    /// Borrowed component batches required that a pointer to the passed in ("adapted") data
+    /// outlives the component batch. Owned component batches on the other hand take ownership by
+    /// allocating a std::vector and moving the data into it. This is typically only required when
+    /// passing in temporary objects into an adapter or non-trivial data conversion is necessary.
+    ///
+    /// By implementing your own adapters for certain component types, you can map your data to
+    /// Rerun types which then can be logged.
+    ///
+    /// TODO(andreas): Point to an example here and in the assert.
+    template <typename TComponent, typename TInput>
+    struct ComponentBatchAdapter {
+        template <typename... Ts>
+        struct NoAdapterFor : std::false_type {};
+
+        static_assert(
+            NoAdapterFor<TComponent, TInput>::value,
+            "ComponentBatchAdapter is not implemented for this type. "
+            "It is implemented for for single components as well as std::vector, std::array, and "
+            "c-arrays of components. "
+            "You can add your own implementation by specializing "
+            "ComponentBatchAdapter<TComponent, T> for a given "
+            "component and your input type T."
+        );
+    };
+
+    /// Type of ownership of the the batch's data.
+    ///
+    /// User access to this is typically only needed for debugging and testing.
+    enum class BatchOwnership {
+        /// The component batch does not own the data and only has a pointer and a
+        /// size.
+        Borrowed,
+
+        /// The component batch owns the data via an std::vector.
+        VectorOwned,
+
+        /// The component batch was moved.
+        /// This could be achieved by other means, but this makes it easiest to follow and
+        /// debug.
+        Moved,
```

Comment:
Please expand on this. Is this a marker for a "ghost" `ComponentBatch` (whose contents have been moved out), or that it contains something that was moved into it?

Reply:
After talking: it is the `ghost` thing.